
# Analysis

```{r alpha, message=F, echo=F, echo=FALSE, warning=FALSE}
library(knitr)
library(knitcitations) 
library(microbiome)
library(phyloseq)
library(ggplot2)
library(reshape2)
library(tidyverse)
library(dada2)
library(dplyr)
library(microbiome)
library(microbiome)
library(phyloseq)
library(vegan)

theme_set(theme_bw(20))
knitr::opts_chunk$set(fig.width=10, fig.height=10, message=FALSE, warning=FALSE)
knitr::opts_chunk$set(fig.path="figure_location/")
opts_chunk$set(dev="CairoPNG")

# Was created with: source("create_phyloseq.R")
phy <- readRDS("data/processed/phyloseq/phy20.1.RDS")

method <- "PCoA"
trans <- "compositional"
distance <- "bray"
index <- "diversity_shannon"
```


## Alpha diversity analysis

Diversity index: `r index`

```{r alpha1, echo=FALSE, message=FALSE, warning=FALSE}
# Estimate alpha diversities
A <- alpha(phy)
```


```{r group_comp, echo=FALSE, echo=FALSE, message=FALSE, fig.width=10, fig.height=6, fig.show="keep", out.width="50%"}
library(Cairo)
df <- meta(phy)
df <- bind_cols(df, A)
df$index <- df[[index]]
pv <- kruskal.test(data = df, index ~ factor(Geographical_location))$p.value
library(ggbeeswarm)
p1 <- ggplot(df, aes(x = Geographical_location, y = index, fill = Geographical_location)) +
        ggbeeswarm::geom_beeswarm() +
	theme(axis.text.x = element_text(angle = 0, hjust = 1)) +
        labs(y = "Shannon Diversity",
	     x = "",
	     title = paste0("Kruskal test p=", round(pv, 2)))

print(p1)
```



## Differential abundance analysis (with Kruskal-Wallis test)

In (jointanalysis.md) it was shown that geographical location has a significant effect.

Here, we investigate individual taxonomic groups in more detail.

Significant (or marginally significant) taxa between geographical locations.

```{r diffab, echo=FALSE, message=FALSE, fig.width=9, fig.height=5, out.width="25%", fig.show="hold"}
location <- meta(phy)$Geographical_location

#Clr-transformation
A <- abundances(microbiome::transform(phy, "clr"))

pv <- apply(A, 1, function (x) {kruskal.test(x ~ location)$p.value});
padj <- p.adjust(pv);

df <- data.frame(taxon = names(padj), padj = padj)
# Add full names for the taxa
full.names <- apply(tax_table(phy), 1, function (x) {paste(x, collapse = "_")})
df$full_name <- full.names[df$taxon]
df <- df %>% filter(padj < 0.25) %>% arrange(padj)
rownames(df) <- NULL

library(knitr)
kable(df)

ps <- microbiome::transform(phy, "clr")
d <- meta(ps)


for (tax in df$taxon) {
  d$taxa <- abundances(ps)[tax, ]
  
  p <- ggplot(d, aes(x = Geographical_location, y = taxa)) +
    geom_boxplot() + 
    ggbeeswarm::geom_beeswarm() +
    theme(axis.text.x = element_text(angle = 360, hjust = 1)) +
    labs(x = "Location", y = "Abundance (CLR)", title = tax)
  print(p)
}

```



```{r heatmaps, echo=FALSE, fig.width=10, fig.height=5, warning=FALSE}

#Initializing data frame
otu_vs_sample <- data.frame(matrix(ncol = length(df$taxon),nrow = length(abundances(ps)[1,])))
colnames(otu_vs_sample) <- df$taxon
rownames(otu_vs_sample) <- colnames(abundances(ps))
otu_vs_sample$location <- NA

#Collecting the data to data frame
for (i in c(1:length(df$taxon))){
  for (j in c(1:length(abundances(ps)[1,]))){
    #Abundances e.g. otu_vs_sample["I1", "OTU237"] <- abundances(ps)["OTU237", I1"]
    otu_vs_sample[rownames(otu_vs_sample)[j],colnames(otu_vs_sample)[i]] <- abundances(ps)[colnames(otu_vs_sample)[i],rownames(otu_vs_sample)[j]]
    #Location information e.g.  otu_vs_sample$location[1] <- d$Geographical_location[d$Sample == "I1"]
    otu_vs_sample$location[j] <- d$Geographical_location[d$Sample == rownames(otu_vs_sample)[j]] 
  }
}

#Rownames to data frame's 1st column
library(data.table)
setDT(otu_vs_sample, keep.rownames = TRUE)[]
colnames(otu_vs_sample)[1] <- "Sample"

#Subsetting data by location
pune_data <- otu_vs_sample %>% filter(location == "Pune")
ahmednagar_data <- otu_vs_sample %>% filter(location == "Ahmednagar")
nashik_data <- otu_vs_sample %>% filter(location == "Nashik")

#Deletes location data
pune_data$location <- NULL
ahmednagar_data$location <- NULL
nashik_data$location <- NULL

#Changing data frames' format
pune_data <- melt(pune_data)
colnames(pune_data) <- c("Sample", "Taxa", "value")

ahmednagar_data <- melt(ahmednagar_data)
colnames(ahmednagar_data) <- c("Sample", "Taxa", "value")

nashik_data <- melt(nashik_data)
colnames(nashik_data) <- c("Sample", "Taxa", "value")

#Creating heatmaps
pune_heat <- heat(pune_data, "Sample","Taxa", "value", legend.text = "Abundance",order.rows = F, 
   order.cols = F, limits= c(-10,10), step = 1) + ggtitle("Pune") + theme(axis.text = element_text(size = 8), axis.text.x = element_blank(), plot.title = element_text(size = 8), legend.title = element_text(size = 8), legend.text = element_text(size = 8), legend.key.height = unit(1.5, "cm"))
 
ahmednagar_heat <- heat(ahmednagar_data, "Sample","Taxa", "value", legend.text = "Abundance",order.rows = F,
  order.cols = F, limits= c(-10,10), step = 1) + ggtitle("Ahmednagar") + theme(axis.text = element_text(size = 8), axis.text.x = element_blank(), plot.title = element_text(size = 8), legend.title = element_text(size = 8), legend.text = element_text(size = 8), legend.key.height = unit(1.5, "cm"))

nashik_heat <- heat(nashik_data, "Sample","Taxa", "value", legend.text = "Abundance",order.rows = F,
  order.cols = F, limits= c(-10,10), step = 1) + ggtitle("Nashik") + theme(axis.text = element_text(size = 8), axis.text.x = element_blank(), plot.title = element_text(size = 8), legend.title = element_text(size = 8), legend.text = element_text(size = 8), legend.key.height = unit(1.5, "cm"))

#Arranging heatmaps to one picture
library(cowplot)
heatmap_grid <- plot_grid(pune_heat + theme(legend.position="none"), ahmednagar_heat + theme(legend.position="none"), nashik_heat + theme(legend.position="none"), ncol = 3, axis = "t", scale = 0.85)

#Displaying the heatmaps and legend
heatmap_grid + draw_grob(get_legend(pune_heat+ theme(legend.position="bottom", legend.key.width = unit(2, "cm"), legend.key.height = unit(0.2, "cm"))), 0.5, -0.38, 0, 1)
```


## Differential abundance analysis 

Differential abundance analysis with DESeq2 method

```{r DESeq2, echo=FALSE, message=FALSE, warning=FALSE ,fig.width=9, fig.height=5, out.width="50%", fig.show="hold"}
library(phyloseq)
library(reshape2)
library(DESeq2)
padj.th <- 0.25
mypseq <- phy

# Remove NAs
group <- "Geographical_location"
keep <- meta(phy)[, group] %>% complete.cases()

df <- meta(mypseq)[keep, ]

mygroup <- "Geographical_location"
df$group <- as.factor(df[[group]] == mygroup)
sample_data(mypseq) <- sample_data(df)
quiet(ds2 <- phyloseq_to_deseq2(mypseq, ~ Geographical_location))
dds <- DESeq(ds2)
res <- results(dds)
df <- as.data.frame(res)
df$taxon <- rownames(df)
#df <- df %>% arrange(desc(abs(log2FoldChange)), padj) %>%
#             filter(padj < padj.th) 
df <- df %>% arrange(padj, desc(abs(log2FoldChange))) %>%
             filter(padj < padj.th) 
df <- subset(df, select= c(log2FoldChange, padj, taxon))
rownames(df) <- NULL

# WHat is the purpose of this one?
#full.names <- apply(tax_table(phy), 1, function (x) {paste(x, collapse = "_")})
# top.taxa <- as.character(df$taxon)
# Add full phylogenetic name for this OTU
df$full_name <- full.names[df$taxon]
library(knitr)
kable(df)


df <- meta(phy)
ps <- microbiome::transform(phy, "clr")
for (tax in top.taxa) {
  df$taxa <- abundances(ps)[tax, ]
  p <- ggplot(df, aes(x = Geographical_location, y = taxa)) +
    geom_boxplot() + 
    ggbeeswarm::geom_beeswarm() +
    theme(axis.text.x = element_text(angle = 360, hjust = 1)) +
    labs(x = "Geographical location", y = "Abundance", title = tax)
  print(p)
}
```